File: 435_Non-overlapping_Intervals/435_1.py
from typing import List

class Solution:
    def eraseOverlapIntervals(self, intervals: List[List[int]]) -> int:
        if not intervals:
            return 0
        intervals.sort(key=lambda x: x[1])
        end = intervals[0][1]
        count = 0
        for i in range(1, len(intervals)):
            if intervals[i][0] < end:
                count += 1
            else:
                end = intervals[i][1]
        return count


if __name__ == "__main__":
    a = [[1,2], [2,3], [3,4], [1,3]]
    s = Solution()
    print(s.eraseOverlapIntervals(a))

File: 435_Non-overlapping_Intervals/435_1.cpp
#include <iostream>
#include <string>
#include <vector>
#include <climits>
#include <algorithm>
#include <cstring>
using namespace std;

bool compare(vector<int> a, vector<int> b) {
    return a[1] < b[1];
}

struct compare2 {
    bool operator() (vector<int> a, vector<int> b) {
        return a[1] < b[1];
    }
} myCompare;

class Solution {
public:
    int eraseOverlapIntervals(vector<vector<int>>& intervals) {
        if (intervals.empty()) return 0;
        sort(intervals.begin(), intervals.end(), myCompare);
        int end = intervals[0][1];
        int count = 0;
        for (int i = 1; i < intervals.size(); ++i) {
            if (intervals[i][0] < end) count++;
            else                       end = intervals[i][1];
        }
        return count;
    }
};


int main() {
    vector<vector<int>> a{{1, 2}, {2, 3}, {3, 4}, {1, 3}};
    Solution s;
    cout << s.eraseOverlapIntervals(a) << endl;
}
